src/service-workers/firebase-messaging/notifications.ts

```diff
@@ -0,0 +1,240 @@
+// Be careful what you import here as it will increase the service worker bundle size
+
+import { get as getFromIndexedDb } from 'idb-keyval'
+import { formatUnits } from '@ethersproject/units' // Increases bundle significantly but unavoidable
+import { getChainsConfig, getBalances, setBaseUrl } from '@safe-global/safe-gateway-typescript-sdk'
+import type { TokenInfo } from '@safe-global/safe-gateway-typescript-sdk'
+import type { ChainInfo } from '@safe-global/safe-gateway-typescript-sdk'
+import type { MessagePayload } from 'firebase/messaging'
+
+import { AppRoutes } from '@/config/routes' // Has no internal imports
+import { isWebhookEvent, WebhookType } from './webhook-types'
+import {
+  getPushNotificationPrefsKey,
+  createPushNotificationPrefsIndexedDb,
+} from '@/services/push-notifications/preferences'
+import { FIREBASE_IS_PRODUCTION } from '@/services/push-notifications/firebase'
+import type { WebhookEvent } from './webhook-types'
+import type { PushNotificationPreferences, PushNotificationPrefsKey } from '@/services/push-notifications/preferences'
+
+export const shouldShowServiceWorkerPushNotification = async (payload: MessagePayload): Promise<boolean> => {
+  if (!isWebhookEvent(payload.data)) {
+    return true
+  }
+
+  const { chainId, address, type } = payload.data
+
+  const key = getPushNotificationPrefsKey(chainId, address)
+  const store = createPushNotificationPrefsIndexedDb()
+
+  const preferencesStore = await getFromIndexedDb<PushNotificationPreferences[PushNotificationPrefsKey]>(
+    key,
+    store,
+  ).catch(() => null)
+
+  if (!preferencesStore) {
+    return false
+  }
+
+  return preferencesStore.preferences[type]
+}
+
+const GATEWAY_URL_PRODUCTION = process.env.NEXT_PUBLIC_GATEWAY_URL_PRODUCTION || 'https://safe-client.safe.global'
+const GATEWAY_URL_STAGING = process.env.NEXT_PUBLIC_GATEWAY_URL_STAGING || 'https://safe-client.staging.5afe.dev'
+
+// localStorage cannot be accessed in service workers so we reference the flag from the environment
+const GATEWAY_URL = FIREBASE_IS_PRODUCTION ? GATEWAY_URL_PRODUCTION : GATEWAY_URL_STAGING
+setBaseUrl(GATEWAY_URL)
```

Comment:
Could you move this to the top of the file?

Reply:
Addressed in a115c14973ce9cf0f7b6af14d7ae633ef6af035d.